W-14927671  Pdk ga branch gr (#464)
mulesoft/docs-gateway: pdk/1.0/modules/ROOT/pages/policies-pdk-configure-features-jwt.adoc

= Using JWT Library Functions
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]
:imagesdir: ../assets/images

NOTE: To view an example policy project that uses PDK's JWT Library, see https://github.com/mulesoft/pdk-custom-policy-examples/blob/main/jwt-validation/README.md[JWT Validation Policy Example^].

> JSON Web Token (JWT) is a URL-secure method of representing claims to be transferred between two parties. The JWT token contains claims encoded in a JSON object as either the payload of a JSON Web Signature (JWS) or as a JSON web encryption (JWE) structure in plain text. This enables the claims to be digitally signed and protected with a message authentication code (MAC). Because the token is signed, you can trust the information and its source.

Reviewer: ```suggestion
JSON Web Token (JWT) is a URL-secure method of representing claims to be transferred between two parties. The JWT token contains claims encoded in a JSON object as either the payload of a JSON Web Signature (JWS) or as a JSON web encryption (JWE) structure in plain text, which enables the claims to be digitally signed and protected with a message authentication code (MAC). Because the token is signed, you can trust the information and its source.
```

* `CustomProvider`: The custom `TokenProvider` that enables you to define a DataWeave expressions to extract the token.

[source,Rust]
----
pub enum TokenProvider {
    #[default]
    BearerProvider,
    CustomProvider {
        expression: Script,
    },
}
----

Both providers implement the `TokenProvision` trait:

[source,Rust]
----
pub trait TokenProvision {
    fn provide(&self, handler: &dyn AttributesBinding) -> Result<String, JWTError>;
}
----

=== Bearer Provider

The following example demonstrates how to extract tokens using the `BearerProvider`:

[source,Rust]
----
use pdk::hl::*;
use pdk::jwt::*;
use pdk::script::HandlerAttributesBinding;
// [...]

async fn filter(
    state: RequestState,
) -> Flow<()> {
    let headers_state = state.into_headers_state().await;

    // Extract token
    let token = TokenProvider::BearerProvider
        .provide(&HandlerAttributesBinding::partial(headers_state.handler()));
// [...]
----

=== Custom Provider

To extract a custom token:

. Define a DataWeave expression in the configuration parameters. For example, the following YAML snippet defines the `tokenExtractor` property:
+
[source,yaml]
----

> ---

Reviewer: ```suggestion

```
I think this line might cause rendering issues, please check in a local build and remove it if the (---) is unnecessary.

[source,Rust]
----
pub fn new(algorithm: SigningAlgorithm, key_length: SigningKeyLength, key: String) -> Result<SignatureValidator, JWTError>

----

PDK supports the following signing algorithms and key lengths:

[%autowidth.spread,cols="a,a"]
|===
| Signing Algorithm | Signing Key Length
| RSA | 256, 384 and 512
| HMAC | 256, 384 and 512
| ES | 256 and 384
|===

The `SigningAlgorithm` enum defines the available signing algorithms:

[source,Rust]
----
pub enum SigningAlgorithm {
    Rsa,
    Hmac,
    Es,
}
----

The `SigningKeyLength` enum defines the available signing key lengths:

[source,Rust]
----
pub enum SigningKeyLength {
    Len256,
    Len384,
    Len512,
}
----

A `SignatureValidator` instance implements the `SignatureValidation` trait. This `SignatureValidation` trait exposes a `validate` function that accepts tokens as `String` parameters and returns a `Result` that contains the parsed `JWTClaims` or a `JWTError` in the event of an error:

[source,Rust]
----
pub trait SignatureValidation {
    fn validate(&self, token: String) -> Result<JWTClaims, JWTError>;
}
----

> The following code example shows how to initialize the the `SignatureValidator` and validate an incoming token signed with a 256 byte HMAC algorithm:

Reviewer: ```suggestion
The following code example shows how to initialize the `SignatureValidator` and validate an incoming token signed with a 256-byte HMAC algorithm:
```

    token_provider: &TokenProvider,
) -> Flow<()> {
    let headers_state = state.into_headers_state().await;

    // Extract token
    let token = token_provider.provide(&HandlerAttributesBinding::partial(headers_state.handler()));

    if token.is_err() {
        return Flow::Break(Response::new(401).with_body("Bearer not found"));
    }

    // Validating signature
    let claims = signature_validator.validate(token.unwrap());

    if claims.is_err() {
        return Flow::Break(Response::new(401).with_body("Invalid token"));
    }
    // [...]
}

#[entrypoint]
async fn configure(launcher: Launcher, Configuration(configuration): Configuration) -> Result<()> {
    let config: Config = serde_json::from_slice(&configuration)?;

    let signature_validator = SignatureValidator::new(
        model::SigningAlgorithm::Hmac,
        model::SigningKeyLength::Len256,
        config.clone().secret,
    )?;

    let token_provider = TokenProvider::CustomProvider { expression: config.token_extractor.clone() };

    launcher
        .launch(on_request(|request| {
            filter(request, &signature_validator, &token_provider)
        }))
        .await?;
    Ok(())
}
----

> Because, the `SignatureValidator` confirms that the policy configurations is correct during initialization, initializing the `SignatureValidator` in the `#[entrypoint]` function throws an early error when the policy is applied rather than when the API receives a request.

Reviewer: ```suggestion
Because, the `SignatureValidator` confirms that the policy configurations are correct during initialization, initializing the `SignatureValidator` in the `#[entrypoint]` function throws an early error when the policy is applied rather than when the API receives a request.
```

After initialization, the policy passes the `SignatureValidator` to the `on_request` wrapped function to perform `validate` method on incoming requests.

== Claim Validation

Because JWT tokens typically contain a set of standard claims that can be validated, the JWT library provides a set of claim validators in the `ClaimValidator` enum:

[source,Rust]
----
pub enum ClaimValidator {
    Aud {
        mandatory: bool,
        audiences: HashSet<String>,
    },
    Nb {
        mandatory: bool,
        current_time: DateTime<Utc>,
    },
    Exp {
        mandatory: bool,
        current_time: DateTime<Utc>,
    },
}
----

> The `mandatory` parameter defines if the claim must be present. If `mandatory: true`, the claim must be present and valid. If `mandatory: false`, the claim is not required. However, if the claim is present, it must be valid. 

Reviewer: ```suggestion
The `mandatory` parameter defines if the claim must be present:

* `mandatory: true`, the claim is required. 
* `mandatory: false`, the claim is not required. 

If the claim is present, it must be valid. 
```

. If the `mandatory` parameter is set to `true` in initialization, it verifies the `exp` JWT claim is included in the incoming token.
. If the `exp` claim is included, it verifies that the `current_time` parameter passed in the `ClaimValidator::Exp` initialization is before the `exp` claim value.

If either verification fails, the validate function returns an error.

The following example shows how to validate expiration in a wrapped function:

[source,rust]
----
use chrono::Utc;
// [...]

async fn filter(
    // [...]
) -> Flow<()> {
    // [...]
    // Validating signature
    let claims = signature_validator.validate(token.unwrap());

    if claims.is_err() {
        return Flow::Break(Response::new(401).with_body("Invalid token"));
    }

    let claims = claims.unwrap();

    // Validating expiration
    let exp_validator = ClaimValidator::Exp {
        mandatory: true,
        current_time: Utc::now(),
    };

    let exp_validation = exp_validator.validate(&claims);

    if exp_validation.is_err() {
        return Flow::Break(Response::new(401).with_body("Expired token"));
    }
    // [...]
}
----

=== Nb (Not Before) Claim Validation

> `ClaimValidator::Nb` is very similar to `ClaimValidator::Exp`. To test validity `ClaimValidator::Nb` compares the `nb` claim value with the parametrized `current_time` to verify if the `current_time` is equal or later than the `nb` claim value.

Reviewer: ```suggestion
`ClaimValidator::Nb` is very similar to `ClaimValidator::Exp`. To test validity `ClaimValidator::Nb` compares the `nb` claim value with the parametrized `current_time` to verify if the `current_time` is equal to or later than the `nb` claim value.
```

    claim_name: String,
    rule: Script,
}
----

The `ScriptClaimRule` has two parameters:

* `claim_name`: The name of the claim the rule validates.
* `rule`: The DataWeave expression that validates the claim logic that has to be executed. 
+
To learn more about defining Dataweave configuration parameters, see xref:policies-pdk-create-schema-definition.adoc#dataweave-expressions[DataWeave Expressions].

The `ScriptClaimValidator` implements the `ScriptClaimValidation` trait that exposes the claim validation method:

[source,rust]
----
pub trait ScriptClaimValidation {
    fn validate(&self, claims: &JWTClaims, handler: &dyn AttributesBinding)
        -> Result<(), JWTError>;
}
----

To instantiate a `ScriptClaimValidator` and use it to perform custom validations:

. Define a DataWeave parameter to act as the custom claim's rule in the policy's schema definition:
+
[source,yaml]
----
# definition/gcl.yaml
# [...]
    customRule:
      type: string
      format: dataweave
      default: "#[vars.claimSet.role=='Member']"
      bindings:
        payload:
          mimeType: text
        attributes: true
        authentication: false
        vars: [claimSet]
# [...]
----

> . Initialize the `ScriptClaimValidator` in the entrypoint function using the configured DataWeave expression a the claim rule and inject the validator into the wrapped function:

Reviewer: ```suggestion
. Initialize the `ScriptClaimValidator` in the entry point function using the configured DataWeave expression and the claim rule and inject the validator into the wrapped function:
```
Please check if this is the intended meaning.

= Using JWT Library Functions
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]
:imagesdir: ../assets/images

NOTE: To view an example policy project that uses PDK's JWT Library, see https://github.com/mulesoft/pdk-custom-policy-examples/blob/main/jwt-validation/README.md[JWT Validation Policy Example^].

JSON Web Token (JWT) is a URL-secure method of representing claims to be transferred between two parties. The JWT token contains claims encoded in a JSON object as either the payload of a JSON Web Signature (JWS) or as a JSON web encryption (JWE) structure in plain text, which enables the claims to be digitally signed and protected with a message authentication code (MAC). Because the token is signed, you can trust the information and its source.

The Flex Gateway Policy Development Kit (PDK) JWT library provides a set of tools to validate the JWT signatures and to extract and validate the claims of all incoming requests.

> == Token extraction

Reviewer 2: For tutorials, "==" level sections need to be imperative tasks.

Author: https://github.com/mulesoft/docs-gateway/pull/504/files

To extract the JWT token, the library provides a `TokenProvider` enum that has two variants:

* `BearerProvider`: The default `TokenProvider` that extracts tokens with the format `Bearer <token>`.
* `CustomProvider`: The custom `TokenProvider` that enables you to define a DataWeave expressions to extract the token.

[source,Rust]
----
pub enum TokenProvider {
    #[default]
    BearerProvider,
    CustomProvider {
        expression: Script,
    },
}
----

Both providers implement the `TokenProvision` trait:

[source,Rust]
----
pub trait TokenProvision {
    fn provide(&self, handler: &dyn AttributesBinding) -> Result<String, JWTError>;
}
----

=== Bearer Provider

The following example demonstrates how to extract tokens using the `BearerProvider`:

[source,Rust]
----
use pdk::hl::*;
use pdk::jwt::*;
use pdk::script::HandlerAttributesBinding;
// [...]

async fn filter(
    state: RequestState,
) -> Flow<()> {
    let headers_state = state.into_headers_state().await;

    // Extract token
    let token = TokenProvider::BearerProvider
        .provide(&HandlerAttributesBinding::partial(headers_state.handler()));
// [...]
----

=== Custom Provider

To extract a custom token:

> . Define a DataWeave expression in the configuration parameters. For example, the following YAML snippet defines the `tokenExtractor` property:

Reviewer 2: It looks like there are two steps required to extract a custom token? But there's quite a bit of information after the second step. Are they all steps? Are only these two required, and everything else is optional?

. Use the following code example to extract the custom JWT token with the defined expression:
+
[source,Rust]
----
use anyhow::Result;
use pdk::hl::*;
use pdk::jwt::*;
use pdk::script::HandlerAttributesBinding;
// [...]

async fn filter(
    state: RequestState,
    token_provider: &TokenProvider,
) -> Flow<()> {
    let headers_state = state.into_headers_state().await;

    // Extract token
    let token = token_provider.provide(&HandlerAttributesBinding::partial(headers_state.handler()));
    // [...]
}

#[entrypoint]
async fn configure(launcher: Launcher, Configuration(configuration): Configuration) -> Result<()> {
    let config: Config = serde_json::from_slice(&configuration)?;

    // Initialize token provider
    let token_provider = TokenProvider::CustomProvider { expression: config.token_extractor.clone() };

    launcher
        .launch(on_request(|request| {
            filter(request, &token_provider)
        }))
        .await?;
    Ok(())
}
----
+
NOTE: The `#[entrypoint]` function initializes `TokenProvider` and then passes is as parameter to the `on_request` wrapped function where the token is extracted.

[[signature-validation]]

> == Signature Validation

Reviewer 2: For tutorials, "==" level sections need to be imperative tasks.

    if claims.is_err() {
        return Flow::Break(Response::new(401).with_body("Invalid token"));
    }
    // [...]
}

#[entrypoint]
async fn configure(launcher: Launcher, Configuration(configuration): Configuration) -> Result<()> {
    let config: Config = serde_json::from_slice(&configuration)?;

    let signature_validator = SignatureValidator::new(
        model::SigningAlgorithm::Hmac,
        model::SigningKeyLength::Len256,
        config.clone().secret,
    )?;

    let token_provider = TokenProvider::CustomProvider { expression: config.token_extractor.clone() };

    launcher
        .launch(on_request(|request| {
            filter(request, &signature_validator, &token_provider)
        }))
        .await?;
    Ok(())
}
----

Because, the `SignatureValidator` confirms that the policy configurations are correct during initialization, initializing the `SignatureValidator` in the `#[entrypoint]` function throws an early error when the policy is applied rather than when the API receives a request.

After initialization, the policy passes the `SignatureValidator` to the `on_request` wrapped function to perform `validate` method on incoming requests.

> == Claim Validation

Reviewer 2: For tutorials, "==" level sections need to be imperative tasks.

Author: https://github.com/mulesoft/docs-gateway/pull/504

    claim_name: String,
    rule: Script,
}
----

The `ScriptClaimRule` has two parameters:

* `claim_name`: The name of the claim the rule validates.
* `rule`: The DataWeave expression that validates the claim logic that has to be executed. 
+
To learn more about defining Dataweave configuration parameters, see xref:policies-pdk-create-schema-definition.adoc#dataweave-expressions[DataWeave Expressions].

The `ScriptClaimValidator` implements the `ScriptClaimValidation` trait that exposes the claim validation method:

[source,rust]
----
pub trait ScriptClaimValidation {
    fn validate(&self, claims: &JWTClaims, handler: &dyn AttributesBinding)
        -> Result<(), JWTError>;
}
----

To instantiate a `ScriptClaimValidator` and use it to perform custom validations:

. Define a DataWeave parameter to act as the custom claim's rule in the policy's schema definition:
+
[source,yaml]
----
# definition/gcl.yaml
# [...]
    customRule:
      type: string
      format: dataweave
      default: "#[vars.claimSet.role=='Member']"
      bindings:
        payload:
          mimeType: text
        attributes: true
        authentication: false
        vars: [claimSet]
# [...]
----

> . Initialize the `ScriptClaimValidator` in the `#[entrypoint]` function using the configured DataWeave expression a the claim rule and inject the validator into the wrapped function:

Reviewer 2: ```suggestion
. Initialize the `ScriptClaimValidator` in the `#[entrypoint]` function using the configured DataWeave expression for the claim rule, and inject the validator into the wrapped function:
```
Is it supposed to be "for"?

    )]
    .to_vec();

    let custom_validator: ScriptClaimValidator =
        ScriptClaimValidator::new(custom_rules, Vec::new());

   // [...]

    launcher
        .launch(on_request(|request| {
            filter(
                request,
                &custom_validator,
                // [...]
            )
        }))
        .await?;
    Ok(())
}
----

. Perform custom validations in the wrapped function by executing the injected validator:
+
[source,rust]
----
async fn filter(
    // [...]
    custom_validator: &ScriptClaimValidator,
) -> Flow<()> {
    // [...]

    let claims = signature_validator.validate(token.unwrap());

    if claims.is_err() {
        return Flow::Break(Response::new(401).with_body("Invalid token"));
    }

    let claims = claims.unwrap();

    // Custom claim validation

    let custom_validation = custom_validator.validate(
        &claims,
        &HandlerAttributesBinding::partial(headers_state.handler()),
    );

    if custom_validation.is_err() {
        return Flow::Break(
            Response::new(400).with_body("Invalid token: Only authenticated customers allowed"),
        );
    }
    
    // [...]
}
----

> == Accessing Claims and Propagating to Headers

Reviewer 2: For tutorials, "==" level sections need to be imperative tasks.

Author: https://github.com/mulesoft/docs-gateway/pull/504
